File: Coursework 2/code/kdl_kine/src/kdl_kine/kdl_kine_solver.py
#!/usr/bin/env python

import rospy
from sensor_msgs.msg import JointState
from kdl_kine.urdf import *

class robot_kinematic(object):
    def __init__(self, base_link, ee_link):

        # Setup the subscribers for the joint states
        self.subscriber_joint_state_ = rospy.Subscriber('/joint_states', JointState, self.joint_state_callback,
                                                        queue_size=5)

        # PyKDL
        self.kine_tree = PyKDL.Tree()
        self.current_pose = PyKDL.Frame()

        (ok, self.kine_tree) = treeFromParam("robot_description")

        self.kine_chain = self.kine_tree.getChain(base_link, ee_link)
        self.NJoints = self.kine_chain.getNrOfJoints()
        self.current_joint_position = PyKDL.JntArray(self.NJoints)
        self.current_joint_velocity = PyKDL.JntArray(self.NJoints)
        self.fk_solver = PyKDL.ChainFkSolverPos_recursive(self.kine_chain)
        self.ik_solver = PyKDL.ChainIkSolverPos_LMA(self.kine_chain)
        self.jac_calc = PyKDL.ChainJntToJacSolver(self.kine_chain)
        self.dyn_solver = PyKDL.ChainDynParam(self.kine_chain, PyKDL.Vector(0, 0, -9.8))

    def joint_state_callback(self, msg):
        # Copies joint position into KDL JntArray
        for i in range(0, self.NJoints):
            self.current_joint_position[i] = msg.position[i]
            # self.current_joint_velocity[i] = msg.velocity[i]

    def forward_kinematics(self, joint):
        #This "joint" is PyKDL.JntArray. The output is a PyKDL.Frame. 
        pose = PyKDL.Frame()
        self.fk_solver.JntToCart(joint, pose)
        return pose

    def get_jacobian(self, joint):
        #This "joint" is PyKDL.JntArray. The output is a PyKDL.Jacobian. 
        jac = PyKDL.Jacobian(self.NJoints)
        self.jac_calc.JntToJac(joint, jac)
        return jac

    def inverse_kinematics_closed(self, desired_pose):
        # This "desired_pose" is PyKDL.Frame. 
        # The output is a PyKDL.JntArray.
        required_joint = PyKDL.JntArray(self.NJoints)
        # Current joint array, desired frame, required joint
        self.ik_solver.CartToJnt(self.current_joint_position, desired_pose, required_joint)
        return required_joint


    def getB(self, joint_val):
        #This "joint_val" is an array. The output is a numpy 2D array. 
        q = PyKDL.JntArray(self.NJoints)
        KDL_B = PyKDL.JntSpaceInertiaMatrix(self.NJoints)

        for i in range(0, self.kine_chain.getNrOfJoints()):
            q[i] = joint_val[i]

        self.dyn_solver.JntToMass(q, KDL_B)
        return KDL_B


    def getC(self, joint_val, joint_vel):
        #This "joint_val" and "joint_vel" are 1D arrays. The output is a JntArray.
        q = PyKDL.JntArray(self.NJoints)
        qdot = PyKDL.JntArray(self.NJoints)
        KDL_C = PyKDL.JntArray(self.NJoints)

        for i in range(0, self.NJoints):
            q[i] = joint_val[i]
            qdot[i] = joint_vel[i]

        self.dyn_solver.JntToCoriolis(q, qdot, KDL_C)

        return KDL_C


    def getG(self, joint_val):
        #This "joint_val" is a 1D array. The output is a JntArray.
        q = PyKDL.JntArray(self.NJoints)
        KDL_G = PyKDL.JntArray(self.NJoints)

        for i in range(0, self.NJoints):
            q[i] = joint_val[i]

        self.dyn_solver.JntToGravity(q, KDL_G)
        return KDL_G

File: Coursework 3/CW3 Code/kdl_kine/src/kdl_kine_solver.cpp
#include <kdl_kine/kdl_kine_solver.h>
#include <string>

void robot_kinematic::init(std::string base_link, std::string ee_link)
{

    subscriber_joint_state = nh.subscribe<sensor_msgs::JointState>("/joint_states", 5, &robot_kinematic::joint_state_callback, this);

    setup_kdl_chain();

    kdl_tree.getChain(base_link, ee_link, kine_chain);

    NrJoints = kine_chain.getNrOfJoints();

    KDL_joint = KDL::JntArray(NrJoints);

    current_joint_position.resize(NrJoints);
    current_joint_velocity.resize(NrJoints);
}

void robot_kinematic::joint_state_callback(const sensor_msgs::JointState::ConstPtr &q)
{
    for (int i = 0; i < NrJoints; i++)
    {
        current_joint_position.data(i) = q->position.at(i);
        current_joint_velocity.data(i) = q->velocity.at(i);
    }
}


void robot_kinematic::setup_kdl_chain() {
    //Take physical parameters from the parameter "robot_description" as defined in the launch file. This will trace back to the urdf file.
    if (!kdl_parser::treeFromParam("robot_description", kdl_tree))
        ROS_ERROR("Failed to construct kdl tree");
}

KDL::Frame robot_kinematic::KDLfkine(KDL::JntArray current_joint_position)
{
    KDL::ChainFkSolverPos_recursive fk_solver = KDL::ChainFkSolverPos_recursive(this->kine_chain);
    KDL::Frame pose;
    fk_solver.JntToCart(current_joint_position, pose, NrJoints);
    return pose;
}

KDL::Jacobian robot_kinematic::KDLjacob(KDL::JntArray current_joint_position)
{
    KDL::ChainJntToJacSolver jac_solver = KDL::ChainJntToJacSolver(this->kine_chain);
    KDL::Jacobian jac = KDL::Jacobian(this->kine_chain.getNrOfJoints());
    jac_solver.JntToJac(current_joint_position, jac);
    return jac;
}

KDL::JntArray robot_kinematic::inverse_kinematics_closed(KDL::Frame desired_pose)
{

    KDL::ChainIkSolverPos_LMA ik_solver = KDL::ChainIkSolverPos_LMA(this->kine_chain);
    KDL::JntArray required_joint = KDL::JntArray(this->kine_chain.getNrOfJoints());
    ik_solver.CartToJnt(this->current_joint_position, desired_pose, required_joint);
    return required_joint;
}

MatrixXd robot_kinematic::getB(KDL::JntArray joint_val)
{
    KDL::JntSpaceInertiaMatrix KDL_B = KDL::JntSpaceInertiaMatrix(NrJoints);
    KDL::ChainDynParam dyn = KDL::ChainDynParam(kine_chain, KDL::Vector(0, 0, -9.8));

    dyn.JntToMass(joint_val, KDL_B);

    return KDL_B.data;
}


VectorXd robot_kinematic::getC(KDL::JntArray joint_val, KDL::JntArray joint_vel)
{


    KDL::JntArray KDL_C = KDL::JntArray(NrJoints);


    KDL::ChainDynParam dyn = KDL::ChainDynParam(kine_chain, KDL::Vector(0, 0, -9.8));


    dyn.JntToCoriolis(joint_val, joint_vel, KDL_C);

    return KDL_C.data;
}


VectorXd robot_kinematic::getG(KDL::JntArray joint_val)
{
    KDL::ChainDynParam dyn = KDL::ChainDynParam(kine_chain, KDL::Vector(0, 0, -9.8));

    KDL::JntArray gravity = KDL::JntArray(NrJoints);

    dyn.JntToGravity(joint_val, gravity);

    return gravity.data;
}
